src/HieDb/Create.hs

```diff
@@ -61,26 +62,56 @@ checkVersion k db@(getConn -> conn) = do
   else
     throwIO $ IncompatibleSchemaVersion dB_VERSION ver
 
+withHieDb' :: FilePath -> ContT a IO HieDb
+withHieDb' fp = do
+  connection <- ContT $ withConnection fp
+  liftIO $ setupHieDb connection
+  let createStatement t = fmap StatementFor (ContT (withStatement connection t))
+      deleteInternalTables = do
+        deletions
+          <- traverse createStatement
+            [ "DELETE FROM refs  WHERE hieFile = ?"
+            , "DELETE FROM decls WHERE hieFile = ?"
+            , "DELETE FROM defs  WHERE hieFile = ?"
+            , "DELETE FROM typerefs WHERE hieFile = ?"
+            , "DELETE FROM mods  WHERE hieFile = ?"
+            , "DELETE FROM exports WHERE hieFile = ?"
+            ]
+        pure $ \fp -> mapM_ (`runStatementFor_` fp) deletions
+  HieDb connection
+    <$> createStatement "INSERT INTO mods VALUES (?,?,?,?,?,?,?)"
+    <*> createStatement "INSERT INTO refs VALUES (?,?,?,?,?,?,?,?,?)"
+    <*> createStatement "INSERT INTO decls VALUES (?,?,?,?,?,?,?)"
+    <*> createStatement "INSERT INTO imports VALUES (?,?,?,?,?,?)"
+    <*> createStatement "INSERT INTO defs VALUES (?,?,?,?,?,?)"
+    <*> createStatement "INSERT INTO exports VALUES (?,?,?,?,?,?,?,?)"
+    <*> createStatement "INSERT INTO typerefs VALUES (?,?,?,?,?,?,?)"
+    <*> createStatement "INSERT INTO typenames(name,mod,unit) VALUES (?,?,?)"
+    <*> createStatement "SELECT id FROM typenames WHERE name = ? AND mod = ? AND unit = ?"
+    <*> deleteInternalTables
+
 {-| Given path to @.hiedb@ file, constructs 'HieDb' and passes it to given function. -}
 withHieDb :: FilePath -> (HieDb -> IO a) -> IO a
-withHieDb fp f = withConnection fp (checkVersion f . HieDb)
+withHieDb fp f = runContT (withHieDb' fp) $ \hiedb -> do
+  checkVersion f hiedb
 
 {-| Given GHC LibDir and path to @.hiedb@ file,
 constructs DynFlags (required for printing info from @.hie@ files)
 and 'HieDb' and passes them to given function.
 -}
 withHieDbAndFlags :: LibDir -> FilePath -> (DynFlags -> HieDb -> IO a) -> IO a
-withHieDbAndFlags libdir fp f = do
+withHieDbAndFlags libdir fp f = runContT (withHieDb' fp) $ \hiedb -> do
   dynFlags <- dynFlagsForPrinting libdir
-  withConnection fp (checkVersion (f dynFlags) . HieDb)
+  f dynFlags hiedb
 
 {-| Initialize database schema for given 'HieDb'.
 -}
-initConn :: HieDb -> IO ()
```

Comment:
We should probably not change the library's api functions without good reason.
E.g. this  functions is exposed in multiple versions of the library (https://hackage-content.haskell.org/package/hiedb-0.7.0.0/docs/HieDb-Create.html#v:initConn) and people using hiedb are probably using it.